source/credentials_provider_imds.c

```diff
@@ -709,12 +729,15 @@ static void s_imds_query_complete(struct aws_credentials_provider_imds_user_data
         (void *)imds_user_data->imds_provider);
 }
 
-static void s_imds_query_error(struct aws_credentials_provider_imds_user_data *imds_user_data) {
+static void s_imds_query_error(struct imds_user_data *imds_user_data) {
     AWS_PRECONDITION(
         imds_user_data->query_state == AWS_IMDS_QS_UNRECOVERABLE_ERROR ||
         imds_user_data->query_state == AWS_IMDS_QS_QUERY_NEVER_CLEARED_STACK);
 
-    imds_user_data->original_callback(NULL, imds_user_data->original_user_data);
+    if (!imds_user_data->callback_invoked) {
```

Comment:
What is making this bool thread safe? is there a lock held above this callback?

Reply:
`imds_user_data` is associated to single credentials request. The workflow is processed by a single thread.

Reply:
it's not all in the same thread on the same connection. It is guaranteed to be sequentially accessed though, and any processor with cache coherency policies (all of them) are fine.